onnxruntime/core/framework/cloud_invoker.cc

```diff
@@ -33,6 +36,122 @@ CloudEndPointInvoker::CloudEndPointInvoker(const CloudEndPointConfig& config,
+                           const InlinedVector<std::string>& input_names,
+                           gsl::span<const OrtValue> ort_inputs,
+                           const InlinedVector<std::string>& output_names,
+                           std::vector<OrtValue>& ort_outputs) const override;
+
+ private:
+  std::string uri_;
+  std::string model_name_;
+};
+
+OpenAIInvoker::OpenAIInvoker(const CloudEndPointConfig& config,
+                             const AllocatorPtr& allocator) : CloudEndPointInvoker(config, allocator) {
+  ReadConfig(kAzureUri, uri_);
+  ReadConfig(kAzureModelName, model_name_);
+}
+
+struct MemoryStruct {
+  char* memory;
+  size_t size;
+};
+
+static size_t
+WriteMemoryCallback(void* contents, size_t size, size_t nmemb, void* userp) {
+  size_t realsize = size * nmemb;
+  struct MemoryStruct* mem = (struct MemoryStruct*)userp;
+
+  char* ptr = (char*)realloc(mem->memory, mem->size + realsize + 1);
+  ORT_ENFORCE(ptr, "not enough memory (realloc returned NULL)");
+
+  mem->memory = ptr;
+  memcpy(&(mem->memory[mem->size]), contents, realsize);
+  mem->size += realsize;
+  mem->memory[mem->size] = 0;
+
+  return realsize;
+}
+
+onnxruntime::Status OpenAIInvoker::Send(const CloudEndPointConfig& run_options,
+                                        const InlinedVector<std::string>& /*input_names*/,
+                                        gsl::span<const OrtValue> ort_inputs,
+                                        const InlinedVector<std::string>& /*output_names*/,
+                                        std::vector<OrtValue>& ort_outputs) const {
+  const auto auth_key_iter = run_options.find(kAzureAuthKey);
+  if (run_options.end() == auth_key_iter || auth_key_iter->second.empty()) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "auth key must be specified for openai client");
+  }
+
+  CURLcode ret{};
+  CURL* curl{};
+  curl_mime* mime1{};
+  struct curl_slist* headers{};
+
+  struct MemoryStruct chunk;
+  chunk.memory = (char*)malloc(1); /* will be grown as needed by the realloc above */
+  chunk.size = 0;
+
+  mime1 = NULL;
+  std::string full_auth = std::string{"Authorization: Bearer "} + auth_key_iter->second;
+  headers = curl_slist_append(headers, full_auth.c_str());
+  headers = curl_slist_append(headers, "Content-Type: multipart/form-data");
+
+  struct curl_httppost* post = NULL;
+  struct curl_httppost* last = NULL;
+  curl_formadd(&post, &last, CURLFORM_COPYNAME, "model", CURLFORM_COPYCONTENTS, model_name_.c_str(), CURLFORM_END);
+  curl_formadd(&post, &last, CURLFORM_COPYNAME, "response_format", CURLFORM_COPYCONTENTS, "text", CURLFORM_END);
+  const auto& tensor = ort_inputs[0].Get<Tensor>();
+  auto data_size = tensor.SizeInBytes();
+  curl_formadd(&post, &last, CURLFORM_COPYNAME, "file", CURLFORM_BUFFER, "non_exist.wav", CURLFORM_BUFFERPTR, tensor.DataRaw(),
+               CURLFORM_BUFFERLENGTH, data_size, CURLFORM_END);
+  curl = curl_easy_init();
```

Comment:
It's better to put this struct to OrtEnv. We should not initialize/deinit it every time for every op invocation.

Reply:
Test especially regarding the curl init, it's pretty lightweight.

Reply:
It is not just about lightweight. It is also about connection reuses across calls.

Reply:
Additionally, this is invoked in the Run call and since Run can be invoked from multiple threads, it can cause issues. Please see doc here https://curl.se/libcurl/c/curl_easy_init.html.

Reply:
Added call to global init with thread-safety ensured.